gc.c

```diff
@@ -1059,17 +1060,67 @@ rb_data_object_check(VALUE klass)
     }
 }
 
+static VALUE typed_data_alloc(VALUE klass, VALUE typed_flag, void *datap, const rb_data_type_t *type, size_t size);
+
+/*
+ * Legacy RData support: RData is implemented as an embedded RTypedData
+ * with the mark/free/data stored in the payload. RDATA(obj)->data still
+ * works because the payload starts at the same offset as RData.data.
+ */
+struct rb_legacy_rdata {
+    void *data;           /* must be first for RDATA(obj)->data compatibility */
+    RUBY_DATA_FUNC dmark;
+    RUBY_DATA_FUNC dfree;
+};
```

Comment:
It is possible to keep the order as previous and make extra embed size 0.
You can obtain the container pointer from the member pointer with `ccan_container_of`.